component/discovery/aws/ec2.go

```diff
@@ -0,0 +1,107 @@
+package aws
+
+import (
+	"errors"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/component/discovery"
+	"github.com/grafana/agent/pkg/flow/rivertypes"
+	promcfg "github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	promaws "github.com/prometheus/prometheus/discovery/aws"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name:    "discovery.aws.ec2",
+		Args:    EC2Arguments{},
+		Exports: discovery.Exports{},
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return NewEC2(opts, args.(EC2Arguments))
+		},
+	})
+}
+
+// EC2Filter is the configuration for filtering EC2 instances.
+type EC2Filter struct {
+	Name   string   `river:"name,attr"`
+	Values []string `yaml:"values,attr"`
+}
+
+// EC2Arguments is the configuration for EC2 based service discovery.
+type EC2Arguments struct {
+	Endpoint        string            `river:"endpoint,attr,optional"`
+	Region          string            `river:"region,attr,optional"`
+	AccessKey       string            `river:"access_key,attr,optional"`
+	SecretKey       rivertypes.Secret `river:"secret_key,attr,optional"`
+	Profile         string            `river:"profile,attr,optional"`
+	RoleARN         string            `river:"role_arn,attr,optional"`
+	RefreshInterval time.Duration     `river:"refresh_interval,attr,optional"`
+	Port            int               `river:"port,attr,optional"`
+	Filters         []*EC2Filter      `river:"filters,block,optional"`
```

Comment:
Blocks should have singular names to feel more natural to write:

```river
filter {
  ...
}

filter {
  ...
}
```

feels less weird to me compared to:

```river
filters {
  ...
}

filters {
  ...
}
```